mitodl/ocw-hugo-themes
# perf: subset icon fonts

# What are the relevant tickets?
Closes https://github.com/mitodl/ocw-hugo-themes/issues/1134


# Description (What does it do?)
This PR:

- Adds `*.codepoints` files for the icons. These are official codepoints for the Material icons. Taken from [here](https://github.com/google/material-design-icons/tree/master/font).
- Adds `*.subset.codepoints` these are the codepoints we want to have in our subsetted font. I discovered 31 icons in total that we use right now.
- Adds a `yarn subset-fonts` command to create subset font files. 
- Updates font URLs to the new subset font assets.
- Adds [docs](https://github.com/mitodl/ocw-hugo-themes/blob/hussaintaj/1134-font-subset/base-theme/assets/fonts/material-design-icons/README.md) about subsetting fonts. I added the main readme doc directly in the fonts directory and added a reference to it in the repo's readme.
- Updates webpack configuration to remove unnecessary files from the output.


This change achieves the following size reduction (on disk) in the icon font assets.

| Asset | Original Size | Subset Size |
| --- | --- | --- |
| MaterialIcons-Regular.ttf | 357 KB | 5 KB |
| MaterialIcons-Regular.woff | 172 KB | 3 KB |
| MaterialIcons-Regular.woff2 | 129 KB | 2 KB |
| MaterialIconsRound-Regular.woff2 | 174 KB | 0.8 KB |

# How can this be tested?

For this PR, you will run two of the following tests:

## Testing Builds

1. Navigate to your local setup of `ocw-hugo-themes`.
2. Switch to the branch `hussaintaj/1134-font-subset`.
3. (Optional) Update `base-theme/assets/css/main.scss` and add the following rule:
    ```scss
    .material-icons,
    .material-icons-round {
      background-color: red !important;
      min-width: 10px !important;
      min-height: 10px !important;
    }
    ```
    > This will mark all icons red, allowing you to spot mistakes easily. 
4. Grab [this course](https://github.mit.edu/ocw-content-rc/ht-test-site) content and [this ocw-www](https://github.mit.edu/ocw-content-rc/ocw-www) content, and place it in your content folder.
5. Run
    ```zsh
    yarn start course ht-test-site
    ```
6. Explore the website at [localhost:3000](http://localhost:3000). Ensure that:
    * All expected icons are visible.
    * Open the network tab and reload the page. The icon font being downloaded should be the `*.subset.*` font. And the size should be less than 10 KB.
    * (Optional, if you followed step 3) You see no empty red boxes.
    > Some learning resource types don't have an icon. They will appear empty. Cross-verify that with RC if in doubt.
7. Run
    ```zsh
    yarn start www
    ```
8. Repeat step 6.
9. Run
    ```zsh
    yarn build /path/to/ocw-content-rc/ht-test-site /path/to/ocw-hugo-projects/ocw-course-v2/config-offline.yaml
    ```
10. Open the file `/path/to/ocw-content-rc/ht-test-site/dist/index.html` in your browser.
11. Repeat step 6.
12. Open the directory `/path/to/ocw-content-rc/ht-test-site/dist/static_shared/fonts/`. Ensure that:
      * You only see 8 fonts. No other files.
      * Material icons have a `.subset.*` extension.
      * Icon fonts have a size of less than 10KB.
13. Run
    ```zsh
    yarn build /path/to/ocw-content-rc/ocw-www /path/to/ocw-hugo-projects/ocw-www/config-offline.yaml
    ```
14. Repeat steps 11 and 12.

## Testing `subset-fonts` Command

Read [these docs](https://github.com/mitodl/ocw-hugo-themes/blob/hussaintaj/1134-font-subset/base-theme/assets/fonts/material-design-icons/README.md). Then follow these steps: 

1. Choose an icon you'd like to add to the subset.
2. Add the icon to the subset according to the instructions in the docs.
3. Update some course content to view your new icon.
4. Run the course with `yarn start course course-id` and ensure that you see your new icon.


## Checklist:
- [x] Add some sort of dev documentation
- [x] Clean build output of any unnecessary files
